Magic attach guide (#2196)
canonical/ubuntu-advantage-client: dev-docs/howtoguides/how_to_use_magic_attach_endpoints.md

# How to use magic attach API endpoints

> 

Reviewer: Let's include a minimum version notice
```suggestion
> **Notice:**
> Minimum version: 27.11
```

> Right now, the pro client provides three distinct endpoints to make it easier to perform

Reviewer: The docs are intended to always be up to date, so I don't think we need the "Right now"
Also `pro` isn't released yet, and these docs will be immediately available, so we should use `ua`.
```suggestion
The UA Client provides three distinct endpoints to make it easier to perform
```

the magic attach flow. They are:

* u.pro.attach.magic.initiate.v1
* u.pro.attach.magic.wait.v1
* u.pro.attach.magic.revoke.v1

We will explain how to use each endpoint and what is the expected output for each of them.

## Initiate endpoint

To start the magic attach flow, we need to create a token for it. The initiate endpoint
will perform exactly that. When you run:

```console

> $ pro api u.pro.attach.magic.initiate.v1

Reviewer: I think we'll replace all of this with `pro` after it's released
```suggestion
$ ua api u.pro.attach.magic.initiate.v1
```

```

It is expected for you to see the following json response:

```json
{
  "_schema_version": "v1",
  "data": {
    "meta": {
       "environment_vars": []}
    },
    "attributes": {
      "expires": "EXPIRE_DATE",
      "expires_in": 10000,
      "token": "MAGIC_ATTACH_TOKEN",
      "user_code": "USER_CODE"
    }
    "type": "MagicAttachInitiate"
  },
  "errors": [],
  "result": "success",
  "version": "PRO CLIENT VERSION",
  "warnings": []
}
```

It is noteworthy here that the `attributes` contain both the `user_code` and `token`. The `user_code`
is the information that will be presented to the user, which it will make possible for the user
to validate the magic attach on the pro portal. Additionally, the `token` information is required
for the other two API endpoints which will be described next.

## Wait endpoint

After we initiate the magic attach procedure, the user must go to the pro portal and validate
the `user_code` it received. Once that is done, a pro token will be generated for the user, allowing

> the attach procedure to begin. The wait endpoint allow us to wait for the user to perform all those

Reviewer: ```suggestion
the attach procedure to begin. The wait endpoint will wait for the user to perform all of those
```

steps on the pro portal. To call it, use:

```console
$ pro api u.pro.attach.magic.initiate.v1 --args magic_token=MAGIC_ATTACH_TOKEN
```

> Note here that the command requires the `token` that was generated in the initiate step. After
> calling this command, we will keep pooling to see if there is any updates on that token. If the

Reviewer: let's make it clearer that the wait command does the polling for you
```suggestion
Note here that the command requires the `token` that was generated in the initiate step.
This command will block and poll the server until there are any updates for that token. If the
```

  "data": {
    "attributes": {
      "contract_id": "CONTRACT_ID",
      "contract_token": "CONTRACT_TOKEN",
      "expires": "EXPIRE_DATE",
      "expires_in": 10000,
      "token": "MAGIC_ATTACH_TOKEN",
      "user_code": "USER_CODE"
    }
    "type": "MagicAttachInitiate"
  },
  "errors": [],
  "result": "success",
  "version": "PRO CLIENT VERSION",
  "warnings": []
}
```

The `contract_token` is the token that can be used to perform an attach operation.

If the provided token is invalid or has expired, we will see the following response:

```json
{
  "_schema_version": "v1",
  "data": {
    "meta": {
      "environment_vars": []
    }
  },
  "errors": [
    {
      "code": "magic-attach-token-error",
      "meta": {},
      "title": "The magic attach token is invalid, has expired or never existed"
    }
  ],
  "result": "failure",
  "version": "PRO CLIENT VERSION",
  "warnings": []
}
```

It is expected that the token will be valid for about 10 minutes. Therefore, we expect the wait
command to keep polling for about that amount of time.

## Revoke

If we want to revoke the token create during the initiate call, we can use the revoke command:

```console

> $ ua pro u.pro.attach.magic.revoke.v1 --args magic_token=MAGIC_ATTACH_TOKEN

Reviewer: ```suggestion
$ ua api u.pro.attach.magic.revoke.v1 --args magic_token=MAGIC_ATTACH_TOKEN
```

      "expires": "EXPIRE_DATE",
      "expires_in": 10000,
      "token": "MAGIC_ATTACH_TOKEN",
      "user_code": "USER_CODE"
    }
    "type": "MagicAttachInitiate"
  },
  "errors": [],
  "result": "success",
  "version": "UA CLIENT VERSION",
  "warnings": []
}
```

It is noteworthy here that the `attributes` contain both the `user_code` and `token`. The `user_code`
is the information that will be presented to the user, which it will make possible for the user
to validate the magic attach on the advantage portal. Additionally, the `token` information is required
for the other two API endpoints which will be described next.

## Wait endpoint

After we initiate the magic attach procedure, the user must go to the advantage portal and validate
the `user_code` it received. Once that is done, a ua token will be generated for the user, allowing
the attach procedure to begin. The wait endpoint will wait for the user to perform all of those
steps on the advantage portal. To call it, use:

```console
$ ua api u.pro.attach.magic.initiate.v1 --args magic_token=MAGIC_ATTACH_TOKEN
```

Note here that the command requires the `token` that was generated in the initiate step. This command

> will block and pool the server until there are any updates for that token. If the

Reviewer 2: `s/pool/poll/`?

response:

```json
{
  "_schema_version": "v1",
  "data": {
    "attributes": {
      "contract_id": "CONTRACT_ID",
      "contract_token": "CONTRACT_TOKEN",
      "expires": "EXPIRE_DATE",
      "expires_in": 10000,
      "token": "MAGIC_ATTACH_TOKEN",
      "user_code": "USER_CODE"
    }
    "type": "MagicAttachInitiate"
  },
  "errors": [],
  "result": "success",
  "version": "UA CLIENT VERSION",
  "warnings": []
}
```

The `contract_token` is the token that can be used to perform an attach operation.

If the provided token is invalid or has expired, we will see the following response:

```json
{
  "_schema_version": "v1",
  "data": {
    "meta": {
      "environment_vars": []
    }
  },
  "errors": [
    {
      "code": "magic-attach-token-error",
      "meta": {},
      "title": "The magic attach token is invalid, has expired or never existed"
    }
  ],
  "result": "failure",
  "version": "UA CLIENT VERSION",
  "warnings": []
}
```

It is expected that the token will be valid for about 10 minutes. Therefore, we expect the wait
command to keep polling for about that amount of time.

## Revoke

> If we want to revoke the token create during the initiate call, we can use the revoke command:

Reviewer 2: `s/create/created/`?
